src/abstractdataframe/reshape.jl

```diff
@@ -399,3 +399,109 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
     res.refs = repeat(res.refs, inner = [v.inner], outer = [v.outer])
     res
 end
+
+
+Base.transpose(::AbstractDataFrame, args...; kwargs...) =
+    MethodError("`transpose` not defined for `AbstractDataFrame`s. Try `permutedims` instead")
+
+"""
+    permutedims(df::AbstractDataFrame, src_namescol::ColumnIndex,
+                dest_namescol::Union{Symbol,AbstractString};
+                makeunique::Bool=false)
+    permutedims(df::AbstractDataFrame, src_namescol::ColumnIndex; makeunique::Bool=false)
+    permutedims(df::AbstractDataFrame; makeunique::Bool=false)
+
+Turn `df` on its side such that rows become columns
+and the column indexed by `src_namescol` becomes the names of new columns.
+In the resulting `DataFrame`,
+The header of `df` will become the first column
+with name specified by `dest_namescol`.
+
+# Arguments
+- `df` : the `AbstractDataFrame`
+- `src_namescol` : the column that will become the new header.
+  This column eltype must be `<: Union{String, Symbol}`.
+  Defaults to first column.
+- `dest_namescol` : the name of the first column in the returned `DataFrame`.
+  Defaults to the same name as `src_namescol`.
```

Comment:
Does this default makes sense in practice? Do we have examples of similar functions in other languages?

Reply:
I do not think we have alternative examples, but it kind of makes sense because such a column is most naturally called `"row_names"` or something like this and after transpose it is still row names (complementing column names).

Reply:
Agree. An alternative way of thinking is that if you're looking at a table of variables that are stored in columns, and you need them stored in rows (my most typical use), the first column name becomes the first row name. Eg

| sample | obs1 | obs2 |
|-- |--|--|
| a | 1.2 | 3.14 |
| b | 42.0 | 1.0 |

becomes

| sample | a | b |
|--|--|--|
| obs1 | 1.2 | 42.0|
| obs2 | 3.14 | 1.0|